read_write.go

```diff
@@ -0,0 +1,95 @@
+package epplib
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"math"
+
+	"github.com/pkg/errors"
+)
+
+// ErrMessageSize represent an error where the incoming message size is bigger than allowed.
+var ErrMessageSize = errors.New("message size exceeds limit")
+
+// MessageReader returns an io.Reader that reads one message according to the
+// message size header. Blocks until a message size header is read.
+// The message limit is in bytes.
+func MessageReader(r io.Reader, msgLimit int64) (io.Reader, error) {
+	// Get the size of the message we are going to read in the future.
+	// https://tools.ietf.org/html/rfc5734#section-4
+	var totalSize uint32
+
+	err := binary.Read(r, binary.BigEndian, &totalSize)
+	if err != nil {
+		return nil, err
+	}
+
+	// The size that the client sent is the size including the bytes that tells
+	// the size so we need to subtract that to get the actual message size.
+	messageSize := int64(totalSize - uint32(binary.Size(totalSize)))
```

Comment:
What happens when `messageSize` becomes negative?

Reply:
We could change the check to `messageSize <= 0 || (msgLimit != 0 && messageSize > msgLimit)` not sure when this would happen though?

Reply:
If the given size is smaller than an uint32.